### PR TITLE
Client Functional Tests

### PR DESCRIPTION
**Pull request by hc-jordenlowe**
Please check out the run-client-functional-tests.sh.  Brian brought up a good point on building the authorization image instead of using the image from docker.  This is so you can test authorization code against it.  Makes sense if we are going to migrate functional tests to this script.

Because I lost some work, some of the code may be formatted poorly or have weird naming.  Point it out and I can fix it.  I thought I got everything, but this has be just ugh.  

Lastly, I will set up a review of this code to show the workflow we had to recreate.  

Fun times!

### CODE REVIEW

**Comment by bsmithvt** [comment]
Perhaps we can make it an option of the `run-client-functional-tests.sh` script to run against a local version of Auth or the version in Docker Hub.

File: Fabric.Authorization.Client.FunctionalTests/FunctionalTestFixture.cs [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+                  Accept = {
+                    new MediaTypeWithQualityHeaderValue("application/json")
+                  }
+                }
+            }).ConfigureAwait(false);
+            return httpResponseMessage;
+        }
+
+        public void Dispose()
+        {
```
Comment (michaelvidal24) [inline Fabric.Authorization.Client.FunctionalTests/FunctionalTestFixture.cs]: Should we dispose the HttpClient here?

File: Fabric.Authorization.Client.FunctionalTests/FunctionalTestHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                return new
+                {
+                    id = FunctionalTestConstants.IdentityTestUser,
+                    name = "Functional Test",
+                    topLevelSecurableItem = new { name = FunctionalTestConstants.IdentityTestUser }
+                }.ToJson();
+            }
+        }
+
+        public static string CreateIdentityUser(string username)
```
Comment (michaelvidal24) [inline Fabric.Authorization.Client.FunctionalTests/FunctionalTestHelpers.cs]: Should we rename this to CreateFunctionalTestClient?

File: Fabric.Authorization.Client.FunctionalTests/FunctionalTestHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                      "fabric/identity.manageresources",
+                      "fabric/authorization.read",
+                      "fabric/authorization.write",
+                      "openid",
+                      "profile"
+                  }
+            }.ToJson();
+        }
+
+        public static FormUrlEncodedContent GetUserAccessToken(string username, string password)
```
Comment (michaelvidal24) [inline Fabric.Authorization.Client.FunctionalTests/FunctionalTestHelpers.cs]: Should we rename this to GetResourceOwnerPasswordPostBody or something similar?

**Comment by michaelvidal24** [comment]
Makes sense to make it an option, so we can run against either the latest release version, or a local version.

File: Fabric.Authorization.API/scripts/run-client-functional-tests.sh [hunk 4]
```diff
@@ -0,0 +1,60 @@
+#!/bin/bash
+
+docker stop authz-functional-authorization
```
[the remark this answers is not in the document]
Comment (hc-jordenlowe) [reply Fabric.Authorization.API/scripts/run-client-functional-tests.sh]: right on

File: Fabric.Authorization.Client.FunctionalTests/FunctionalTestFixture.cs [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+
+        public FunctionalTestFixture()
+        {
+            this.GetParameters();
+            this._client = new HttpClient();
+            this._client.BaseAddress = new Uri(this.args.AuthorizationBaseUrl);
+            this.BaseUrl = this._client.BaseAddress;
+
+            Task<string> tokenForInstaller = this.GetAccessTokenForInstaller(this.args.InstallerClientSecret);
+            tokenForInstaller.Wait();
```
[the remark this answers is not in the document]
Comment (hc-jordenlowe) [reply Fabric.Authorization.Client.FunctionalTests/FunctionalTestFixture.cs]: this did have ConfigureAwait(false), stupid compiler changes

[the remark this answers is not in the document]
Comment (hc-jordenlowe) [reply Fabric.Authorization.Client.FunctionalTests/FunctionalTestFixture.cs]: hmm, maybe not

**Pull request merged**